src/include/optimizer/absexpr_expression.h

```diff
@@ -0,0 +1,183 @@
+#include <vector>
+
+namespace peloton {
+namespace optimizer {
+
+// (TODO): rethink the AbsExpr_Container/Expression approach in comparion to abstract
+// Most of the core rule/optimizer code relies on the concept of an Operator /
+// OperatorExpression and the interface that the two functions respectively expose.
+//
+// The annoying part is that an AbstractExpression blends together an Operator
+// and OperatorExpression. Second part, the AbstractExpression does not export the
+// correct interface that the rest of the system depends on.
+//
+// As an extreme level of simplification (sort of hacky), an AbsExpr_Container is
+// analogous to Operator and wraps a single AbstractExpression node. AbsExpr_Expression
+// is analogous to OperatorExpression.
+//
+// AbsExpr_Container does *not* handle memory correctly w.r.t internal instantiations
+// from Rule transformation. This is since Peloton itself mixes unique_ptrs and
+// hands out raw pointers which makes adding a shared_ptr here extremely problematic.
+// terrier uses only shared_ptr when dealing with AbstractExpression trees.
+
+class AbsExpr_Container {
+ public:
+  AbsExpr_Container();
+
+  AbsExpr_Container(const expression::AbstractExpression *expr) {
+    node = expr;
+  }
+
+  // Return operator type
+  ExpressionType GetType() const {
+    if (IsDefined()) {
+      return node->GetExpressionType();
+    }
+    return ExpressionType::INVALID;
+  }
+
+  const expression::AbstractExpression *GetExpr() const {
+    return node;
+  }
+
+  // Operator contains Logical node
+  bool IsLogical() const {
+    return true;
+  }
+
+  // Operator contains Physical node
+  bool IsPhysical() const {
+    return false;
+  }
+
+  std::string GetName() const {
+    if (IsDefined()) {
+      return node->GetExpressionName();
+    }
+
+    return "Undefined";
+  }
+
+  hash_t Hash() const {
+    if (IsDefined()) {
+      return node->Hash();
+    }
+    return 0;
+  }
+
+  bool operator==(const AbsExpr_Container &r) {
+    if (IsDefined() && r.IsDefined()) {
+      // (TODO): need a better way to determine deep equality
+
+      // NOTE:
+      // Without proper equality determinations, the groups will
+      // not be assigned correctly. Arguably, terrier does this
+      // better because a blind ExactlyEquals on different types
+      // of ConstantValueExpression under Peloton will crash!
+
+      // For now, just return (false).
+      // I don't anticipate this will affect correctness, just
+      // performance, since duplicate trees will have to evaluated
+      // over and over again, rather than being able to "borrow"
+      // a previous tree's rewrite.
+      //
+      // Probably not worth to create a "validator" since porting
+      // this to terrier anyways (?). == does not check Value
+      // so it's broken. ExactlyEqual requires precondition checking.
+      return false;
+    } else if (!IsDefined() && !r.IsDefined()) {
+      return true;
+    }
+    return false;
+  }
+
```

Comment:
This may not be a huge deal since, as you said, the == op doesn't affect correctness but do you intend to implement the == op for rewrites? If so, does there exists an == op for AbstractExpression you can use?

Reply:
I believe Terrier currently only has the notion of logical equality for abstract expressions.